File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int mini = 1, maxi = 1, res = INT_MIN;
        
        for(int i: nums) {
            if(i==0) {
                mini=maxi=1;
                res=max(res, 0);
                continue;
            }
            
            int temp = i*maxi;
            maxi = max({temp, i*mini, i});
            mini = min({temp, i*mini, i});
            
            res = max(res, maxi);
        }
        return res;
    }
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.go
func maxProduct(nums []int) int {
    mini, maxi, res := 1, 1, math.MinInt32
    
    for _, i := range nums {
        if i==0 {
            mini=1
            maxi=1
            res = max(res, 0)
            continue
        }
        
        temp := maxi*i
        maxi = max(temp, max(mini*i, i))
        mini = min(temp, min(mini*i, i))
        res = max(res, maxi)
    }
    
    return res
}